packages/api/src/controllers/cdn-data.ts

```diff
@@ -0,0 +1,272 @@
+  async (req, res) => {
+    const { region } = req.params;
+    if (!region) {
+      res.status(400);
+      return res.end();
+    }
+    const [docs] = await db.cdnUsageLast.find({ region: req.params.region });
+    if (!docs?.length) {
+      res.status(204);
+      return res.end();
+    }
+    res.status(200);
+    res.json(docs[0]);
+  }
+);
+
+app.post(
+  "/",
+  //   authMiddleware({}),
+  // validatePost("stream"),
+  async (req, res) => {
+    // if (!req.body || !req.body.name) {
+    //     res.status(422);
+    //     return res.json({
+    //         errors: ["missing name"],
+    //     });
+    // }
+    const usersCache = new Map();
+    const getUser = async (playbackId: string): Promise<User | null> => {
+      if (usersCache.has(playbackId)) {
+        return usersCache.get(playbackId);
+      }
+      const stream = await db.stream.getByPlaybackId(playbackId);
+      if (!stream) {
+        logger.error(`Can't find stream for playbackId=${playbackId}`);
+        return null;
+      }
+      const user = await db.user.get(stream.userId);
+      if (!user) {
+        logger.error(`Can't find user for playbackId=${playbackId}`);
+        return null;
+      }
+      usersCache.set(playbackId, user);
+      return user;
+    };
+
+    const start = Date.now();
+    const dataAr = req.body as Array<SendData>;
+    console.log("Got data: ", JSON.stringify(dataAr, null, 2));
+    if (!Array.isArray(dataAr)) {
+      res.status(400);
+      res.end();
+    }
+    for (const data of dataAr) {
+      const hour = new Date(data.date * 1000).toUTCString();
+      if (!data.data?.length) {
+        continue;
+      }
+      // aggregate stream_ids into playback_ids
+      for (const row of data.data) {
+        // console.log(`===> checking row `, row)
+        if (row.playback_id) {
+          const user = await getUser(row.playback_id);
+          if (user) {
+            row.user_id = user.id;
+            row.user_email = user.email;
+          }
+        }
+        if (row.stream_id) {
+          // find playbackId by streamId
+          let stream = await db.stream.get(row.stream_id);
+          if (!stream) {
+            logger.error(`Invalid stream_id=${row.stream_id} in hour=${hour}`);
+            continue;
+          }
+          if (!stream.playbackId && stream.parentId) {
+            stream = await db.stream.get(stream.parentId);
+          }
+          if (!stream) {
+            logger.error(
+              `Invalid stream_id=${stream.parentId} in hour=${hour}`
+            );
+            continue;
+          }
+          if (!stream.playbackId) {
+            logger.error(
+              `For stream_id=${stream.parentId} in hour=${hour} can't find playbackId.`
+            );
+            continue;
+          }
+          const playbackRow = data.data.find(
+            (row) => row.playback_id === stream.playbackId
+          );
+          if (playbackRow) {
+            playbackRow.count += row.count;
```

Comment:
Curious, what are the cases when there is a row coming from CDN data that only has a `stream_id` and the `playback_id` is on a separate value? Would this catch like API calls for the stream info?